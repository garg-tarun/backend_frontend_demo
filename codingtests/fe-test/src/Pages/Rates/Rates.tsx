import { useState, useEffect, useRef } from "react";
import DropDown from "../../Components/DropDown";
import ProgressBar from "../../Components/ProgressBar";
import Loader from "../../Components/Loader";
import TextInput from "../../Components/TextInput";
import ErrorDialog from "../../Components/ErrorDialog";
import { useAnimationFrame } from "../../Hooks/useAnimationFrame";
import classes from "./Rates.module.css";
import CountryData from "../../Libs/Countries.json";
import countryToCurrency from "../../Libs/CountryCurrency.json";

let countries = CountryData.CountryCodes;

const Rates = () => {
  const [fromCurrency, setFromCurrency] = useState("AU");
  const [toCurrency, setToCurrency] = useState("US");

  const [exchangeRate, setExchangeRate] = useState(0.7456);
  const [progression, setProgression] = useState(0);
  const [loading, setLoading] = useState(false);

  const [amount, setAmount] = useState<number | "">("");
  const [error, setError] = useState<string | null>(null);
  const [inputValue, setInput] = useState("");

  const controllerRef = useRef<AbortController | null>(null);

  const Flag = ({ code }: { code: string }) => (
    <img
      alt={code || ""}
      src={`/img/flags/${code || ""}.svg`}
      width="20px"
      className={classes.flag}
    />
  );

  const handleChange = (raw: string) => {
    // Accept only digits (0-9)
    if (!/^\d*$/.test(raw)) {
      setError("Only digits (0-9) are allowed");
      return;
    }
    setInput(raw);
    if (raw === "") {
      setAmount("");
    } else {
      setAmount(parseInt(raw, 10));
    }
  };

  // Dismiss error after 5s
  useEffect(() => {
    if (error) {
      const timer = setTimeout(() => setError(null), 5000);
      return () => clearTimeout(timer);
    }
  }, [error]);

  // Markup calculation
  const markUpPercent = 0.5;
  const markUpRate = (markUpPercent * exchangeRate) / 100;
  const trueAmount = amount !== "" ? amount * exchangeRate : 0;
  const markedUpAmount =
    amount !== "" ? amount * (exchangeRate - markUpRate) : 0;

  const sellCurrencyCode =
    countryToCurrency[fromCurrency as keyof typeof countryToCurrency];
  const buyCurrencyCode =
    countryToCurrency[toCurrency as keyof typeof countryToCurrency];

  const sendRequest = async (signal?: AbortSignal) => {
    const baseUrl = "https://rates.staging.api.paytron.com/rate/public";
    const params = new URLSearchParams({
      sellCurrency: sellCurrencyCode,
      buyCurrency: buyCurrencyCode,
    });
    const url = `${baseUrl}?${params.toString()}`;
    const response = await fetch(url, signal ? { signal } : undefined);

    if (!response.ok) {
      const errorData = await response.json();
      const errorText = errorData.detail ?? "Unknown error";
      throw new Error(errorText);
    }
    const data = await response.json();
    return data.retailRate;
  };

  const fetchData = async () => {
    if (controllerRef.current) {
      controllerRef.current.abort();
    }

    const controller = new AbortController();
    controllerRef.current = controller;

    setLoading(true);
    setError(null);

    try {
      const retailRate = await sendRequest(controller.signal);
      setExchangeRate(retailRate);
    } catch (err) {
      if ((err as any).name !== "AbortError") {
        setExchangeRate(0);
        setError(err instanceof Error ? err.message : "Unknown error");
      }
    } finally {
      setLoading(false);
    }
  };

  // Fetch excahnge rate if currency is changed
  useEffect(() => {
    if (!fromCurrency || !toCurrency) {
      setExchangeRate(0);
      return;
    }
    fetchData();
    // clean up previous fetch request
    return () => controllerRef.current?.abort();
  }, [fromCurrency, toCurrency]);

  // Demo progress bar moving :)
  useAnimationFrame(!loading, (deltaTime) => {
    setProgression((prevState) => {
      if (prevState > 0.998) {
        fetchData();
        return 0;
      }
      return (prevState + deltaTime * 0.0001) % 1;
    });
  });

  return (
    <div className={classes.container}>
      <div className={classes.content}>
        <div className={classes.heading}>Currency Conversion</div>

        <div className={classes.rowWrapper}>
          <div>
            <DropDown
              leftIcon={<Flag code={fromCurrency} />}
              label={"From"}
              selected={
                countryToCurrency[
                  fromCurrency as keyof typeof countryToCurrency
                ]
              }
              options={countries.map(({ code }) => ({
                option:
                  countryToCurrency[code as keyof typeof countryToCurrency],
                key: code,
                icon: <Flag code={code} />,
              }))}
              setSelected={(key: string) => {
                setFromCurrency(key);
              }}
              style={{ marginRight: "20px" }}
            />
          </div>

          <div className={classes.exchangeWrapper}>
            <div className={classes.transferIcon}>
              <img src="/img/icons/Transfer.svg" alt="Transfer icon" />
            </div>

            <div className={classes.rate}>{exchangeRate}</div>
          </div>

          <div>
            <DropDown
              leftIcon={<Flag code={toCurrency} />}
              label={"To"}
              selected={
                countryToCurrency[toCurrency as keyof typeof countryToCurrency]
              }
              options={countries.map(({ code }) => ({
                option:
                  countryToCurrency[code as keyof typeof countryToCurrency],
                key: code,
                icon: <Flag code={code} />,
              }))}
              setSelected={(key: string) => {
                setToCurrency(key);
              }}
              style={{ marginLeft: "20px" }}
            />
          </div>
        </div>
        <div className={classes.rowWrapper}>
          <TextInput
            id="transfer-amount"
            label="Transfer amount"
            value={inputValue}
            onChange={handleChange}
            placeholder="e.g. 100"
          />
          {error && (
            <ErrorDialog
              message={error}
              onClose={() => setError("")}
              autoCloseDuration={5000}
            />
          )}
        </div>

        {amount !== "" && (
          <div style={{ marginTop: "1rem" }}>
            <p>
              <strong>True Amount:</strong> {trueAmount.toFixed(4)}{" "}
              {buyCurrencyCode}
            </p>
            <p>
              <strong>Markup Amount:</strong> {markedUpAmount.toFixed(4)}{" "}
              {buyCurrencyCode}
            </p>
          </div>
        )}

        <ProgressBar
          progress={progression}
          animationClass={loading ? classes.slow : ""}
          style={{ marginTop: "20px" }}
        />

        {loading && (
          <div className={classes.loaderWrapper}>
            <Loader width={"25px"} height={"25px"} />
          </div>
        )}
      </div>
    </div>
  );
};

export default Rates;
