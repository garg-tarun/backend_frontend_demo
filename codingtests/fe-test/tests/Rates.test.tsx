import {
  render,
  screen,
  fireEvent,
  within,
  waitFor,
} from "@testing-library/react";
import { vi } from "vitest";
import styles from "../src/Pages/Rates/Rates.module.css";
import { describe, expect, it } from "vitest";
import Rates from "../src/Pages/Rates";
import {
  verifyFlagElement,
  selectCurrencyUsingDropdown,
} from "./utils/helpers";

beforeEach(() => {
  vi.stubGlobal(
    "fetch",
    vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ retailRate: 1.54 }),
      })
    )
  );
});

afterEach(() => {
  vi.restoreAllMocks();
});

describe("Rates Page Test", () => {
  const mockSetSelected = vi.fn();
  const fromCurrency = "AU";
  const ToCurrency = "US";

  it("Verify default page rendering", async () => {
    render(<Rates />);
    const title = screen.getByText("Currency Conversion");
    expect(title).toBeInTheDocument();

    const label = screen.getByText("From");
    expect(label).toBeInTheDocument();

    verifyFlagElement(fromCurrency);
    verifyFlagElement(ToCurrency);

    expect(screen.getByLabelText("Transfer amount")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("e.g. 100")).toBeInTheDocument();
  });

  it("Verify page rendering after from/to currency change", async () => {
    render(<Rates />);

    // change from/to currency using dropdown
    const fromLabel = "From";
    const fromCurrencyCode = "SGD";
    selectCurrencyUsingDropdown(fromLabel, fromCurrencyCode);

    const toLabel = "To";
    const toCurrencyCode = "INR";
    selectCurrencyUsingDropdown(toLabel, toCurrencyCode);

    // validate flag after from/to currency change
    const newFromCurrency = "SG";
    const newToCurrency = "IN";
    verifyFlagElement(newFromCurrency);
    verifyFlagElement(newToCurrency);

    // wait for the rate to update (rerender via state)
    await waitFor(() => expect(screen.getByText(/1\.54/)).toBeInTheDocument());

    const input = screen.getByPlaceholderText("e.g. 100");
    fireEvent.change(input, { target: { value: "100" } });
    expect(input).toHaveValue("100");

    await waitFor(() =>
      expect(screen.getByText(/True Amount:/i)).toBeInTheDocument()
    );

    const trueAmountElm = screen.getByText(/True Amount:/i).parentElement!;
    const markupAmountElm = screen.getByText(/Markup Amount:/i).parentElement!;
    expect(trueAmountElm.textContent).toBe("True Amount: 154.0000 INR");
    expect(markupAmountElm.textContent).toBe("Markup Amount: 153.2300 INR");
  });

  it("matches snapshot with default rendering", () => {
    const { container } = render(<Rates />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot after changing from/to currency", () => {
    const { container } = render(<Rates />);
    // change from/to currency using dropdown
    const fromLabel = "From";
    const fromCurrencyCode = "SGD";
    selectCurrencyUsingDropdown(fromLabel, fromCurrencyCode);

    const toLabel = "To";
    const toCurrencyCode = "INR";
    selectCurrencyUsingDropdown(toLabel, toCurrencyCode);
    expect(container).toMatchSnapshot();
  });
});
