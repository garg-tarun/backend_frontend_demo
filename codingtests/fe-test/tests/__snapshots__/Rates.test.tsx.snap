// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rates Page Test > matches snapshot after changing from/to currency 1`] = `
<div>
  <div
    class="_container_3590bb"
  >
    <div
      class="_content_3590bb"
    >
      <div
        class="_heading_3590bb"
      >
        Currency Conversion
      </div>
      <div
        class="_rowWrapper_3590bb"
      >
        <div>
          <div
            class="_container_5d938f undefined"
            style="margin-right: 20px;"
          >
            <span>
              From
            </span>
            <button
              class="_dropdown_5d938f"
            >
              <img
                alt="SG"
                class="_flag_3590bb"
                src="/img/flags/SG.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
              >
                SGD
              </span>
              <div
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="_exchangeWrapper_3590bb"
        >
          <div
            class="_transferIcon_3590bb"
          >
            <img
              alt="Transfer icon"
              src="/img/icons/Transfer.svg"
            />
          </div>
          <div
            class="_rate_3590bb"
          >
            0.7456
          </div>
        </div>
        <div>
          <div
            class="_container_5d938f undefined"
            style="margin-left: 20px;"
          >
            <span>
              To
            </span>
            <button
              class="_dropdown_5d938f"
            >
              <img
                alt="IN"
                class="_flag_3590bb"
                src="/img/flags/IN.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
              >
                INR
              </span>
              <div
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="_rowWrapper_3590bb"
      >
        <div
          style="margin-bottom: 1rem; margin-top: 2rem;"
        >
          <label
            for="transfer-amount"
          >
            Transfer amount
          </label>
          <br />
          <input
            id="transfer-amount"
            inputmode="numeric"
            pattern="-?\\d*"
            placeholder="e.g. 100"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="_base_a40bf7"
        style="margin-top: 20px;"
      >
        <div
          class="_progress_a40bf7 _slow_3590bb"
          style="transform: translate(-0px);"
        />
      </div>
      <div
        class="_loaderWrapper_3590bb"
      >
        <div
          class="_loader_20fc81"
          style="width: 25px; height: 25px;"
        >
          <svg
            fill="none"
            viewBox="0 0 19 19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="loader-circle"
              cx="9.5"
              cy="9.5"
              r="7.5"
              stroke="#8e96aa"
              stroke-opacity="0.5"
              stroke-width="2.5"
            />
            <path
              class="loader-focus"
              d="M9.5 17C5.35786 17 2 13.6421 2 9.5"
              stroke="#457ced"
              stroke-width="2.5"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Rates Page Test > matches snapshot with default rendering 1`] = `
<div>
  <div
    class="_container_3590bb"
  >
    <div
      class="_content_3590bb"
    >
      <div
        class="_heading_3590bb"
      >
        Currency Conversion
      </div>
      <div
        class="_rowWrapper_3590bb"
      >
        <div>
          <div
            class="_container_5d938f undefined"
            style="margin-right: 20px;"
          >
            <span>
              From
            </span>
            <button
              class="_dropdown_5d938f"
            >
              <img
                alt="AU"
                class="_flag_3590bb"
                src="/img/flags/AU.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
              >
                AUD
              </span>
              <div
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="_exchangeWrapper_3590bb"
        >
          <div
            class="_transferIcon_3590bb"
          >
            <img
              alt="Transfer icon"
              src="/img/icons/Transfer.svg"
            />
          </div>
          <div
            class="_rate_3590bb"
          >
            0.7456
          </div>
        </div>
        <div>
          <div
            class="_container_5d938f undefined"
            style="margin-left: 20px;"
          >
            <span>
              To
            </span>
            <button
              class="_dropdown_5d938f"
            >
              <img
                alt="US"
                class="_flag_3590bb"
                src="/img/flags/US.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
              >
                USD
              </span>
              <div
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="_rowWrapper_3590bb"
      >
        <div
          style="margin-bottom: 1rem; margin-top: 2rem;"
        >
          <label
            for="transfer-amount"
          >
            Transfer amount
          </label>
          <br />
          <input
            id="transfer-amount"
            inputmode="numeric"
            pattern="-?\\d*"
            placeholder="e.g. 100"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="_base_a40bf7"
        style="margin-top: 20px;"
      >
        <div
          class="_progress_a40bf7 _slow_3590bb"
          style="transform: translate(-0px);"
        />
      </div>
      <div
        class="_loaderWrapper_3590bb"
      >
        <div
          class="_loader_20fc81"
          style="width: 25px; height: 25px;"
        >
          <svg
            fill="none"
            viewBox="0 0 19 19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="loader-circle"
              cx="9.5"
              cy="9.5"
              r="7.5"
              stroke="#8e96aa"
              stroke-opacity="0.5"
              stroke-width="2.5"
            />
            <path
              class="loader-focus"
              d="M9.5 17C5.35786 17 2 13.6421 2 9.5"
              stroke="#457ced"
              stroke-width="2.5"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
